File: tests/wasm/test_andean_oscillator.js
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { loadTestData, assertArrayClose } from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

function manualAndean(open, close, length = 50, signalLength = 9) {
    const bull = new Float64Array(open.length);
    const bear = new Float64Array(open.length);
    const signal = new Float64Array(open.length);
    bull.fill(NaN);
    bear.fill(NaN);
    signal.fill(NaN);

    const alpha = 2 / (length + 1);
    const signalAlpha = 2 / (signalLength + 1);
    let initialized = false;
    let up1 = NaN;
    let up2 = NaN;
    let dn1 = NaN;
    let dn2 = NaN;
    let sig = NaN;

    for (let i = 0; i < open.length; i++) {
        const o = open[i];
        const c = close[i];
        if (!Number.isFinite(o) || !Number.isFinite(c)) {
            continue;
        }

        const c2 = c * c;
        const o2 = o * o;

        if (!initialized) {
            up1 = c;
            up2 = c2;
            dn1 = c;
            dn2 = c2;
            sig = 0;
            bull[i] = 0;
            bear[i] = 0;
            signal[i] = 0;
            initialized = true;
            continue;
        }

        up1 = Math.max(c, o, up1 - (up1 - c) * alpha);
        up2 = Math.max(c2, o2, up2 - (up2 - c2) * alpha);
        dn1 = Math.min(c, o, dn1 + (c - dn1) * alpha);
        dn2 = Math.min(c2, o2, dn2 + (c2 - dn2) * alpha);

        bull[i] = Math.sqrt(Math.max(0, dn2 - dn1 * dn1));
        bear[i] = Math.sqrt(Math.max(0, up2 - up1 * up1));
        const src = Math.max(bull[i], bear[i]);
        sig = signalAlpha * src + (1 - signalAlpha) * sig;
        signal[i] = sig;
    }

    return { bull, bear, signal };
}

test.before(async () => {
    const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
    const importPath = process.platform === 'win32'
        ? 'file:///' + wasmPath.replace(/\\/g, '/')
        : wasmPath;
    wasm = await import(importPath);
    testData = loadTestData();
});

test('andean_oscillator_js matches manual reference', () => {
    const open = new Float64Array(testData.open.slice(0, 160));
    const close = new Float64Array(testData.close.slice(0, 160));
    const result = wasm.andean_oscillator_js(open, close, 50, 9);
    const expected = manualAndean(open, close, 50, 9);
    assertArrayClose(result.bull, expected.bull, 1e-12, 'bull mismatch');
    assertArrayClose(result.bear, expected.bear, 1e-12, 'bear mismatch');
    assertArrayClose(result.signal, expected.signal, 1e-12, 'signal mismatch');
});

test('andean_oscillator_batch_js first row matches single', () => {
    const open = new Float64Array(testData.open.slice(0, 128));
    const close = new Float64Array(testData.close.slice(0, 128));
    const batch = wasm.andean_oscillator_batch_js(open, close, {
        length_range: [50, 52, 2],
        signal_length_range: [9, 11, 2],
    });
    const single = wasm.andean_oscillator_js(open, close, 50, 9);
    assert.strictEqual(batch.rows, 4);
    assert.strictEqual(batch.cols, 128);
    assert.deepStrictEqual(Array.from(batch.lengths), [50, 50, 52, 52]);
    assert.deepStrictEqual(Array.from(batch.signal_lengths), [9, 11, 9, 11]);
    assertArrayClose(batch.bull.slice(0, 128), single.bull, 1e-12, 'batch bull mismatch');
    assertArrayClose(batch.bear.slice(0, 128), single.bear, 1e-12, 'batch bear mismatch');
    assertArrayClose(batch.signal.slice(0, 128), single.signal, 1e-12, 'batch signal mismatch');
});

test('andean_oscillator_into pointer path matches safe API', () => {
    const n = 144;
    const open = new Float64Array(testData.open.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));
    const safe = wasm.andean_oscillator_js(open, close, 50, 9);
    const openPtr = wasm.allocate_f64_array(n);
    const closePtr = wasm.allocate_f64_array(n);
    const outBullPtr = wasm.andean_oscillator_alloc(n);
    const outBearPtr = wasm.andean_oscillator_alloc(n);
    const outSignalPtr = wasm.andean_oscillator_alloc(n);
    try {
        wasm.write_f64_array(openPtr, open);
        wasm.write_f64_array(closePtr, close);
        wasm.andean_oscillator_into(openPtr, closePtr, outBullPtr, outBearPtr, outSignalPtr, n, 50, 9);
        const bull = wasm.read_f64_array(outBullPtr, n);
        const bear = wasm.read_f64_array(outBearPtr, n);
        const signal = wasm.read_f64_array(outSignalPtr, n);
        assertArrayClose(bull, safe.bull, 1e-12, 'pointer bull mismatch');
        assertArrayClose(bear, safe.bear, 1e-12, 'pointer bear mismatch');
        assertArrayClose(signal, safe.signal, 1e-12, 'pointer signal mismatch');
    } finally {
        wasm.deallocate_f64_array(openPtr);
        wasm.deallocate_f64_array(closePtr);
        wasm.andean_oscillator_free(outBullPtr, n);
        wasm.andean_oscillator_free(outBearPtr, n);
        wasm.andean_oscillator_free(outSignalPtr, n);
    }
});

test('andean_oscillator_batch_into returns row count', () => {
    const n = 128;
    const rows = 4;
    const open = new Float64Array(testData.open.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));
    const openPtr = wasm.allocate_f64_array(n);
    const closePtr = wasm.allocate_f64_array(n);
    const outBullPtr = wasm.andean_oscillator_alloc(rows * n);
    const outBearPtr = wasm.andean_oscillator_alloc(rows * n);
    const outSignalPtr = wasm.andean_oscillator_alloc(rows * n);
    try {
        wasm.write_f64_array(openPtr, open);
        wasm.write_f64_array(closePtr, close);
        const returnedRows = wasm.andean_oscillator_batch_into(
            openPtr,
            closePtr,
            outBullPtr,
            outBearPtr,
            outSignalPtr,
            n,
            50,
            52,
            2,
            9,
            11,
            2,
        );
        assert.strictEqual(returnedRows, rows);
    } finally {
        wasm.deallocate_f64_array(openPtr);
        wasm.deallocate_f64_array(closePtr);
        wasm.andean_oscillator_free(outBullPtr, rows * n);
        wasm.andean_oscillator_free(outBearPtr, rows * n);
        wasm.andean_oscillator_free(outSignalPtr, rows * n);
    }
});

test('andean_oscillator_js rejects invalid length', () => {
    const open = new Float64Array(testData.open.slice(0, 64));
    const close = new Float64Array(testData.close.slice(0, 64));
    assert.throws(
        () => wasm.andean_oscillator_js(open, close, 0, 9),
        /Invalid length/,
    );
});

File: tests/python/test_andean_oscillator.py
import numpy as np
import pytest

from test_utils import load_test_data

try:
    import vector_ta as mp
except ImportError:
    try:
        import vector_ta as mp
    except ImportError:
        pytest.skip(
            "Python module not built. Run 'maturin develop --features python' first",
            allow_module_level=True,
        )


def manual_andean(open_, close, length=50, signal_length=9):
    bull = np.full(open_.shape[0], np.nan, dtype=np.float64)
    bear = np.full(open_.shape[0], np.nan, dtype=np.float64)
    signal = np.full(open_.shape[0], np.nan, dtype=np.float64)
    alpha = 2.0 / (length + 1.0)
    signal_alpha = 2.0 / (signal_length + 1.0)

    initialized = False
    up1 = up2 = dn1 = dn2 = sig = np.nan

    for i in range(open_.shape[0]):
        o = open_[i]
        c = close[i]
        if not np.isfinite(o) or not np.isfinite(c):
            continue

        c2 = c * c
        o2 = o * o

        if not initialized:
            up1 = c
            up2 = c2
            dn1 = c
            dn2 = c2
            sig = 0.0
            bull[i] = 0.0
            bear[i] = 0.0
            signal[i] = 0.0
            initialized = True
            continue

        up1 = max(c, o, up1 - (up1 - c) * alpha)
        up2 = max(c2, o2, up2 - (up2 - c2) * alpha)
        dn1 = min(c, o, dn1 + (c - dn1) * alpha)
        dn2 = min(c2, o2, dn2 + (c2 - dn2) * alpha)

        bull[i] = np.sqrt(max(0.0, dn2 - dn1 * dn1))
        bear[i] = np.sqrt(max(0.0, up2 - up1 * up1))
        src = max(bull[i], bear[i])
        sig = signal_alpha * src + (1.0 - signal_alpha) * sig
        signal[i] = sig

    return bull, bear, signal


class TestAndeanOscillator:
    def test_manual_reference_matches_binding(self):
        data = load_test_data()
        open_ = np.asarray(data["open"][:160], dtype=np.float64)
        close = np.asarray(data["close"][:160], dtype=np.float64)
        expected_bull, expected_bear, expected_signal = manual_andean(open_, close, 50, 9)
        result = mp.andean_oscillator(open_, close, length=50, signal_length=9)
        np.testing.assert_allclose(result["bull"], expected_bull, rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(result["bear"], expected_bear, rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(
            result["signal"], expected_signal, rtol=0.0, atol=1e-12, equal_nan=True
        )

    def test_stream_matches_batch(self):
        data = load_test_data()
        open_ = np.asarray(data["open"][:144], dtype=np.float64)
        close = np.asarray(data["close"][:144], dtype=np.float64)
        batch = mp.andean_oscillator(open_, close, length=50, signal_length=9)
        stream = mp.AndeanOscillatorStream(50, 9)
        streamed = np.array([stream.update(float(o), float(c)) for o, c in zip(open_, close)], dtype=np.float64)
        np.testing.assert_allclose(streamed[:, 0], batch["bull"], rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(streamed[:, 1], batch["bear"], rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(streamed[:, 2], batch["signal"], rtol=0.0, atol=1e-12, equal_nan=True)

    def test_batch_metadata_and_first_row_match_single(self):
        data = load_test_data()
        open_ = np.asarray(data["open"][:128], dtype=np.float64)
        close = np.asarray(data["close"][:128], dtype=np.float64)
        result = mp.andean_oscillator_batch(
            open_,
            close,
            length_range=(50, 52, 2),
            signal_length_range=(9, 11, 2),
        )
        assert result["rows"] == 4
        assert result["cols"] == open_.shape[0]
        np.testing.assert_allclose(result["lengths"], np.array([50, 50, 52, 52]), rtol=0.0, atol=0.0)
        np.testing.assert_allclose(
            result["signal_lengths"], np.array([9, 11, 9, 11]), rtol=0.0, atol=0.0
        )
        single = mp.andean_oscillator(open_, close, length=50, signal_length=9)
        np.testing.assert_allclose(result["bull"][0], single["bull"], rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(result["bear"][0], single["bear"], rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(
            result["signal"][0], single["signal"], rtol=0.0, atol=1e-12, equal_nan=True
        )

    def test_invalid_parameters(self):
        data = load_test_data()
        open_ = np.asarray(data["open"][:64], dtype=np.float64)
        close = np.asarray(data["close"][:64], dtype=np.float64)
        with pytest.raises(ValueError, match="Invalid length"):
            mp.andean_oscillator(open_, close, length=0, signal_length=9)
        with pytest.raises(ValueError, match="Invalid signal_length"):
            mp.andean_oscillator(open_, close, length=50, signal_length=0)
